File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/K._Достопримечательности/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <utility>
#include <limits>


std::vector<int> dijkstra(int& source, std::unordered_map<int, std::vector<std::pair<int, int>>>& adjList, int& n) {
    std::vector<int> dist (n + 1, -1);
    dist[source] = 0;
    std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<std::pair<int, int>>> minHeap;
    minHeap.push({0, source});
    while (minHeap.size() > 0) {
        auto [weight, vertex] = minHeap.top();
        minHeap.pop();
        for (std::pair<int, int>& edge : adjList[vertex]) {
            int amount = weight + edge.second;
            if ((dist[edge.first] == -1) || (amount < dist[edge.first])) {
                dist[edge.first] = amount;
                minHeap.push({amount, edge.first});
            }
        }
    }
    return dist;
}


void solution() {
    /*
    Time Complexity O(V(VlogE))
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::vector<std::pair<int, int>>> adjList;
    for (int i = 0; i < m; ++i) {
        int source, destination, weight;
        std::cin >> source >> destination >> weight;
        adjList[source].push_back({destination, weight});
        adjList[destination].push_back({source, weight});
    }
    for (int vertex = 1; vertex <= n; ++vertex) {
        std::vector<int> shortestPathTree = dijkstra(vertex, adjList, n);
        for (int idx = 1; idx <= n; ++idx) {
            if (idx > 1) std::cout << " ";
            std::cout << shortestPathTree[idx];
        }
        std::cout << "\n";
    }
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/K._Достопримечательности/solution.py
import sys
import heapq


def dijkstra(source: int, adj_list: dict[int, list[int]], n: int) -> list[int]:
    dist: list[int] = [-1] * (n + 1)
    dist[source] = 0
    min_heap: list[tuple[int, int]] = [(0, source)]
    while min_heap:
        weight, vertex = heapq.heappop(min_heap)
        for neighbor, cost in adj_list.get(vertex, []):
            amount: int = weight + cost
            if (dist[neighbor] == -1) or (amount < dist[neighbor]):
                heapq.heappush(min_heap, (amount, neighbor))
                dist[neighbor] = amount
    return dist


def solution() -> None:
    """
    Time Complexity O(V(VlogE))
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(m):
        source, destination, weight = map(int, sys.stdin.readline().rstrip().split())
        if source not in adj_list: adj_list[source] = []
        adj_list[source].append((destination, weight))
        if destination not in adj_list: adj_list[destination] = []
        adj_list[destination].append((source, weight))
    for vertex in range(1, n + 1):
        shortest_path_tree: list[int] = dijkstra(vertex, adj_list, n)
        for idx in range(1, n + 1):
            if idx > 1: sys.stdout.write(' ')
            sys.stdout.write(str(shortest_path_tree[idx]))
        sys.stdout.write('\n')


if __name__ == '__main__':
    solution()